### PR TITLE
Suggest using slice when encountering `let x = ""[..];`

### LINKED ISSUE #26319: A couple of confusing diagnostics with &str slices
https://play.rust-lang.org/?gist=089ca2d7214f6761051c&version=stable

``` rust
let s = "abc";
let t = if true { s[..2] } else { s };       // confusing error message
let u: &str = if true { s[..2] } else { s }; // clear error message
let v = s[..2];                              // strange error message
let w: &str = s[..2];                        // clear error message
```

These are four attempts to take a slice of a `&str`, in two pairs. They are all wrong because indexing a `&str` returns a `str`, so you need an extra `&`. That's not the point. The point is without type annotations, the error diagnostics seem like they could use some improvement.

Especially the first one -- it sounds like it is saying my expression ought to be a `str`, but it already is. I think it's actually saying the associated type `Index::Output` ought to be a `str`, but the way the diagnostic is written, it sounds backwards.

Can either of these diagnostics be improved?

### PR DESCRIPTION
**Pull request by estebank**
Fix #26319.

### CODE REVIEW

**Comment by eddyb** [comment]
I don't think the index type matters, we should just not check it and just emit the suggestion.

File: src/test/ui/suggestions/str-array-assignment.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn main() {
+  let s = "abc";
+  let t = if true { s[..2] } else { s };
```
Comment (oli-obk) [inline src/test/ui/suggestions/str-array-assignment.rs]: You need to add `//~ ERROR` comments. Ui tests work like compile-fail now

File: src/librustc/traits/error_reporting.rs [hunk 2]
```diff
@@ -821,6 +823,33 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         err.emit();
     }
 
+    /// When encountering an assignment of an unsized trait, like `let x = ""[..];`, provide a
```
Comment (oli-obk) [inline src/librustc/traits/error_reporting.rs]: Why just for assignments and not all the cases in the test?

Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: The other cases in the test go through type check instead. I'll see if I can expand #46256 to also suggest in these cases. I included all the cases originally mentioned in order to make sure we don't regress in any of the cases (without noticing).

File: src/test/ui/suggestions/str-array-assignment.rs (hunk 1 above)
Comment (estebank) [reply src/test/ui/suggestions/str-array-assignment.rs]: Done.

**Comment by estebank** [comment]
@eddyb I believe you were right so removed the index type check.

**Comment by eddyb** [comment]
r? @nikomatsakis I still feel this is much more specific than it should be, but I don't have any good suggestions, other than perhaps that `&` isn't called "slice-ing" or "taking a slice".
`a[..]` does the slicing, `&` just borrows that slice. You can also have `a[..].slice_method()`.

File: src/test/ui/suggestions/str-array-assignment.stderr [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected str, found &str
+   |
+   = note: expected type `str`
+              found type `&str`
+
+error[E0308]: mismatched types
+  --> $DIR/str-array-assignment.rs:17:27
+   |
+11 | fn main() { //~ NOTE expected `()` because of default return type
+   |           - expected `()` because of default return type
```
Comment (arielb1) [inline src/test/ui/suggestions/str-array-assignment.stderr]: this looks dubious. Could you look into this?

Comment (estebank) [reply src/test/ui/suggestions/str-array-assignment.stderr]: Yeah, I noticed that too. Still digging.

**Comment by arielb1** [comment]
This feels hacky to me, but I don't see any better way of implementing this (the rhs of locals is an lvalue, so checks based on it being an rvalue won't quite work).

btw, you should be checking that the problem type is a slice or string - if the return type is a struct or trait object, you don't want the error message to mention "trait objects".

File: src/test/ui/suggestions/str-array-assignment.stderr (hunk 3 above)
Comment (arielb1) [reply src/test/ui/suggestions/str-array-assignment.stderr]: it's also already the case on nightly. Feel free to ignore this and open an issue.

**Comment by arielb1** [comment]
OTOH, having an `Index` implementation that returns trait objects is a bit weird - that would mean someone does some kind of indirection somewhere. But I think "consider borrowing here" would look better than "consider borrowing the slice" (also because `str` is not a "slice").

**Comment by estebank** [comment]
@arielb1 reworded.

File: src/test/ui/suggestions/str-array-assignment.stderr (hunk 3 above)
Comment (estebank) [reply src/test/ui/suggestions/str-array-assignment.stderr]: #46302

**Pull request merged**